module4/lessons/Week4/SecurityConcerns.md

```diff
@@ -0,0 +1,187 @@
+
+<img src="/assets/images/module4/week4/hashImage.png" style="width: 80%;">
+
+[From Wikipedia](https://en.wikipedia.org/wiki/Hash_function): “A hash function is any function that can be used to map data of arbitrary size onto data of a fixed size”.
+
+In the image above, each name is mapped to a number.  Where two names are assigned to the same number we have a **collision** (this is bad).  In this example, every name is matched to a two digit number; there are only so many two digit numbers, so it is highly likely that we will have a collision.  In production-level hash functions, data is assigned to a much longer sequence of numbers and letters.
+
+There are a number of [common hash function types](https://en.wikipedia.org/wiki/List_of_hash_functions) that are used in modern web development.  The one we will use today is SHA256.
+
+<section class='call-to-action' markdown='1'>
+
+Create a new console application and put this code in the `program.cs` file.  Run it a few times, and then annotate what each line of code. You will likely need to do some research in order to figure out what each line is doing!
+
+```c#
+HashAlgorithm sha = SHA256.Create();
+
+string firstInput = "Hello, World!";
+Console.WriteLine("Input: " + firstInput);
+byte[] firstInputBytes = Encoding.ASCII.GetBytes(firstInput);
+byte[] firstInputDigested = sha.ComputeHash(firstInputBytes);
+StringBuilder firstInputBuilder = new StringBuilder();
+foreach (byte b in firstInputDigested)
+{
+    Console.Write(b + ", ");
+    firstInputBuilder.Append(b.ToString("x2"));
+}
+Console.WriteLine();
+Console.WriteLine("Digest: " + firstInputBuilder);
+
+Console.WriteLine("\n\n");
+
+var secondInput = "Howdy";
+Console.WriteLine("Input: " + secondInput);
+byte[] secondInputBytes = Encoding.ASCII.GetBytes(secondInput);
+byte[] secondInputDigested = sha.ComputeHash(secondInputBytes);
+StringBuilder secondInputBuilder = new StringBuilder();
+foreach (byte b in secondInputDigested)
+{
+    Console.Write(b + ", ");
+    secondInputBuilder.Append(b.ToString("x2"));
+}
+Console.WriteLine();
+Console.WriteLine("Digest: " + secondInputBuilder);
+```
+
+Be ready to share your annotations!
+
+🌶️If you have time, are there ways that we could clean up this code?
+
+</section>
+
+Hashing data is a good way to securely store data, but it is not fool-proof.  Let's imagine hack into my application and are able to retrieve all my users’ hashed passwords. You find that the account with username `boss@example.com`` has this hashed password:
+
+```
+16e6b8af370f0d6b6e67c65bb6f4499c01a974f1dcd14a5f9429c6931531e6c4
+```
+
+Based on the length of the digest you guess it’s a SHA256. You know that some users, particularly bosses, are lazy and they do dumb things like re-use their 4-digit ATM pin for their password. But the application required a password of eight digits, so they might have repeated the pin.
```

Comment:
I like the idea of this example, but I don't totally follow how we are very close to cracking this user's account. How would we know their ATM pin? What do you think about switching up the example to something more like brute force trying all of the common passwords 12341234, 12345678, etc by running them through SHA256 and seeing what they give us and if that matches the Boss's password? 

That would also reinforce the key point that same input will always give you the same output with the same hash function, at GA we called that the golden rule of hash functions or something like that 😆

Reply:
I think that example would also make it more clear why salting is helpful.

Reply:
I guess by very close, I meant we could reasonably brute force it.  We used to run this as part of the security lesson in BE, and actually had students crack the pin with these constraints.  It takes about 10 minutes 🙃.  I'll re-work this a bit to be more clear; I don't know if we have the time during the lesson to break the pin, but it could be a fun exercise for students if they are interested.